File: src/honeypots/vnc.js
'use strict';

const { TcpHoneypot } = require('./base');

class VncHoneypot extends TcpHoneypot {
  constructor(cfg, logger) {
    super({ name: 'vnc', port: cfg.port }, logger);
    this.cfg = cfg;
  }

  onConnection(socket, meta) {
    let stage = 'version';
    let challenge = null;
    const send = (b) => { try { socket.write(b); } catch { /* noop */ } };
    send(Buffer.from(this.cfg.banner || 'RFB 003.008\n'));

    socket.on('data', (data) => {
      if (stage === 'version') {
        this.logger.log({
          service: 'vnc', type: 'client_version',
          sessionId: meta.sessionId,
          remoteIp: meta.remoteIp, remotePort: meta.remotePort, localPort: meta.localPort,
          details: { version: data.toString('utf8').trim() }
        });
        // Offer VNC Authentication only
        send(Buffer.from([0x01, 0x02]));
        stage = 'sec';
        return;
      }
      if (stage === 'sec') {
        if (data[0] === 0x02) {
          challenge = require('crypto').randomBytes(16);
          send(challenge);
          stage = 'challenge';
        } else {
          try { socket.end(); } catch { /* noop */ }
        }
        return;
      }
      if (stage === 'challenge') {
        this.logger.log({
          service: 'vnc', type: 'auth_attempt',
          sessionId: meta.sessionId,
          remoteIp: meta.remoteIp, remotePort: meta.remotePort, localPort: meta.localPort,
          details: {
            challengeHex: challenge?.toString('hex') || null,
            responseHex: data.slice(0, 16).toString('hex')
          }
        });
        send(Buffer.from([0x00, 0x00, 0x00, 0x01])); // failed
        try { socket.end(); } catch { /* noop */ }
      }
    });
  }
}

module.exports = { create: (cfg, logger) => new VncHoneypot(cfg, logger) };

File: go-honeypot/server/internal/honeypots/vnc.go
package honeypots

import (
	"context"
	cryptorand "crypto/rand"
	"encoding/hex"
	"net"
	"strings"

	"github.com/example/honeypot/internal/config"
	"github.com/example/honeypot/internal/eventlog"
)

func NewVNC(cfg config.Service, store *eventlog.Store) *TCP {
	banner := cfg.Banner
	if banner == "" {
		banner = "RFB 003.008\n"
	}
	return NewTCP("vnc", cfg, store, func(ctx context.Context, conn net.Conn, sessionID string, meta ConnMeta) {
		_, _ = conn.Write([]byte(banner))
		buf := make([]byte, 16)
		n, err := conn.Read(buf)
		if err != nil {
			return
		}
		store.Log(eventlog.Event{
			Service: "vnc", Type: "client_version", SessionID: sessionID,
			RemoteIP: meta.RemoteIP, RemotePort: meta.RemotePort, LocalPort: meta.LocalPort,
			Details: map[string]any{"version": strings.TrimSpace(string(buf[:n]))},
		})
		// Offer VNC Authentication only
		_, _ = conn.Write([]byte{0x01, 0x02})
		sel := make([]byte, 1)
		if _, err := conn.Read(sel); err != nil || sel[0] != 0x02 {
			return
		}
		challenge := make([]byte, 16)
		_, _ = cryptorand.Read(challenge)
		_, _ = conn.Write(challenge)
		resp := make([]byte, 16)
		if _, err := conn.Read(resp); err != nil {
			return
		}
		store.Log(eventlog.Event{
			Service: "vnc", Type: "auth_attempt", SessionID: sessionID,
			RemoteIP: meta.RemoteIP, RemotePort: meta.RemotePort, LocalPort: meta.LocalPort,
			Details: map[string]any{
				"challenge": hex.EncodeToString(challenge),
				"response":  hex.EncodeToString(resp),
			},
		})
		_, _ = conn.Write([]byte{0x00, 0x00, 0x00, 0x01})
		_ = ctx
	})
}
